From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Yive <6853318+Yive@users.noreply.github.com>
Date: Tue, 6 May 2025 22:14:55 -0700
Subject: [PATCH] Pluto: Expose Direction$Plane's faces

The JMH benchmark of this patch can be found in SunBox's `CachedEnumValuesForLoop`

Original license: GPL-3.0-only
Original project: https://github.com/Yive/Pluto

diff --git a/net/minecraft/core/Direction.java b/net/minecraft/core/Direction.java
index e6b14aa7f9bd3003428c4e5756b9daf6a14d1a49..42d1d41f0f818f3f3526ce33c30027041752e6b3 100644
--- a/net/minecraft/core/Direction.java
+++ b/net/minecraft/core/Direction.java
@@ -591,7 +591,7 @@ public enum Direction implements Directional, StringRepresentable, ca.spottedlea
         HORIZONTAL(new Direction[]{Direction.NORTH, Direction.EAST, Direction.SOUTH, Direction.WEST}, new Direction.Axis[]{Direction.Axis.X, Direction.Axis.Z}),
         VERTICAL(new Direction[]{Direction.UP, Direction.DOWN}, new Direction.Axis[]{Direction.Axis.Y});
 
-        private final Direction[] faces;
+        public final Direction[] faces; // Pluto - Expose Direction$Plane's faces - private -> public
         private final Direction.Axis[] axis;
 
         Plane(final Direction[] faces, final Direction.Axis[] axis) {
diff --git a/net/minecraft/data/worldgen/features/VegetationFeatures.java b/net/minecraft/data/worldgen/features/VegetationFeatures.java
index bc700545aa552e843e3e860f1ff383599e1963e9..02521579ae7962578063a66a464aee7889d06988 100644
--- a/net/minecraft/data/worldgen/features/VegetationFeatures.java
+++ b/net/minecraft/data/worldgen/features/VegetationFeatures.java
@@ -493,7 +493,7 @@ public class VegetationFeatures {
         WeightedList.Builder<BlockState> segmentedBlockBuild = WeightedList.builder();
 
         for (int amount = minState; amount <= maxState; amount++) {
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 segmentedBlockBuild.add(block.defaultBlockState().setValue(amountProperty, amount).setValue(directionProperty, direction), 1);
             }
         }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 8b544ddf12b98b46afebd4763b44ac50ef0840d2..5beb82565f7eb2a0e99eb2be1e844bfbd2219c33 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1311,7 +1311,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                         // We only need to check blocks that are taller than the minimum step height
                         if (org.purpurmc.purpur.PurpurConfig.smoothSnowAccumulationStep > 0 && currentLayers >= org.purpurmc.purpur.PurpurConfig.smoothSnowAccumulationStep) {
                             int layersValueMin = currentLayers - org.purpurmc.purpur.PurpurConfig.smoothSnowAccumulationStep;
-                            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                 BlockPos blockPosNeighbor = topPos.relative(direction);
                                 BlockState blockStateNeighbor = this.getBlockState(blockPosNeighbor);
                                 if (blockStateNeighbor.is(Blocks.SNOW)) {
diff --git a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
index 818d7882a4d7001a9d8f90516b7368e89a5689a0..c336c47a026f0906a4e4a4cb6dac86f0de016adb 100644
--- a/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
+++ b/net/minecraft/world/entity/ai/behavior/PrepareRamNearestTarget.java
@@ -145,7 +145,7 @@ public class PrepareRamNearestTarget<E extends PathfinderMob> extends Behavior<E
         List<BlockPos> possibleRamPositions = Lists.newArrayList();
         BlockPos.MutableBlockPos walkablePosFurthestAwayFromTarget = targetPos.mutable();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             walkablePosFurthestAwayFromTarget.set(targetPos);
 
             for (int distance = 0; distance < this.maxRamDistance; distance++) {
diff --git a/net/minecraft/world/entity/ai/behavior/TryFindLandNearLiquid.java b/net/minecraft/world/entity/ai/behavior/TryFindLandNearLiquid.java
index d873c8725cc86a3b3270c140729c23e737b53638..844cab0eab41e3bb4883fd8aa94684fbc7fa4f8c 100644
--- a/net/minecraft/world/entity/ai/behavior/TryFindLandNearLiquid.java
+++ b/net/minecraft/world/entity/ai/behavior/TryFindLandNearLiquid.java
@@ -39,7 +39,7 @@ public class TryFindLandNearLiquid {
                                         return false;
                                     }
 
-                                    for (Direction direction : Direction.Plane.HORIZONTAL) {
+                                    for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                         testPos.setWithOffset(pos, direction);
                                         if (level.getBlockState(testPos).isAir()
                                             && level.getBlockState(testPos.move(Direction.DOWN)).getFluidState().is(fluidTag)) {
diff --git a/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java b/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
index 5ec0d04a0983cf6da1c10728ed36ca1ff9fd70ec..586e54778cf9a6a7cb09c81f4d881f092f182ea6 100644
--- a/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
+++ b/net/minecraft/world/entity/ai/behavior/TryLaySpawnOnFluidNearLand.java
@@ -23,7 +23,7 @@ public class TryLaySpawnOnFluidNearLand {
                         if (!body.isInWater() && body.onGround()) {
                             BlockPos belowPos = body.blockPosition().below();
 
-                            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                 BlockPos relativePos = belowPos.relative(direction);
                                 if (level.getBlockState(relativePos).getCollisionShape(level, relativePos).getFaceShape(Direction.UP).isEmpty()
                                     && (
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/AbstractThrownPotion.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/AbstractThrownPotion.java
index e241312fd7484fa1d867b86828c785a466ff5d82..b535db7f6e4b75903c1576f7bf980ebbde33e075 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/AbstractThrownPotion.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/AbstractThrownPotion.java
@@ -60,7 +60,7 @@ public abstract class AbstractThrownPotion extends ThrowableItemProjectile {
                 this.douseFire(blockEffectPos);
                 this.douseFire(blockEffectPos.relative(hitDirection.getOpposite()));
 
-                for (Direction direction : Direction.Plane.HORIZONTAL) {
+                for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                     this.douseFire(blockEffectPos.relative(direction));
                 }
             }
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 96204ee26dd7e65ed4df6392088fdba53058fb94..da3c7f75838b3160eef9c774d130810d0f68786f 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -2162,7 +2162,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     public abstract Scoreboard getScoreboard();
 
     public void updateNeighbourForOutputSignal(final BlockPos pos, final Block changedBlock) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos relativePos = pos.relative(direction);
             if (this.hasChunkAt(relativePos)) {
                 BlockState state = this.getBlockState(relativePos);
diff --git a/net/minecraft/world/level/block/CactusBlock.java b/net/minecraft/world/level/block/CactusBlock.java
index e2f183dc52cf69153fdcf692442e182cfecaec31..85bf28f79251ba07c7d564f8ba7c022221a0540d 100644
--- a/net/minecraft/world/level/block/CactusBlock.java
+++ b/net/minecraft/world/level/block/CactusBlock.java
@@ -130,7 +130,7 @@ public class CactusBlock extends Block implements BonemealableBlock { // Purpur
     }
     protected boolean canSurvive(final LevelReader level, final BlockPos pos) {
         // Pluto end - Check if the cactus can even survive being placed
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockState neighbor = level.getBlockState(pos.relative(direction));
             if ((level.getWorldBorder().world == null || level.getWorldBorder().world.purpurConfig.cactusBreaksFromSolidNeighbors) && neighbor.isSolid() || level.getFluidState(pos.relative(direction)).is(FluidTags.LAVA)) { // Purpur - Cactus breaks from solid neighbors config
                 return false;
diff --git a/net/minecraft/world/level/block/ChorusFlowerBlock.java b/net/minecraft/world/level/block/ChorusFlowerBlock.java
index 4b2f1868ebd07b97a4900ea55aa31226a2d2e263..7b0b0fca1952f73a9126933011a714b34e0bef63 100644
--- a/net/minecraft/world/level/block/ChorusFlowerBlock.java
+++ b/net/minecraft/world/level/block/ChorusFlowerBlock.java
@@ -133,7 +133,7 @@ public class ChorusFlowerBlock extends Block {
     }
 
     private static boolean allNeighborsEmpty(final LevelReader level, final BlockPos pos, final @Nullable Direction ignore) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (direction != ignore && !level.isEmptyBlock(pos.relative(direction))) {
                 return false;
             }
@@ -170,7 +170,7 @@ public class ChorusFlowerBlock extends Block {
 
             boolean oneNeighbor = false;
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 BlockState neighbor = level.getBlockState(pos.relative(direction));
                 if (neighbor.is(this.plant)) {
                     if (oneNeighbor) {
diff --git a/net/minecraft/world/level/block/ChorusPlantBlock.java b/net/minecraft/world/level/block/ChorusPlantBlock.java
index 9bb7a2a33d96fcc9b05aaa58fb7dac61c1e90e13..933fafc91f8e368651c0581e583439b4b586287b 100644
--- a/net/minecraft/world/level/block/ChorusPlantBlock.java
+++ b/net/minecraft/world/level/block/ChorusPlantBlock.java
@@ -88,7 +88,7 @@ public class ChorusPlantBlock extends PipeBlock {
         BlockState belowState = level.getBlockState(pos.below());
         boolean blockAboveOrBelow = !level.getBlockState(pos.above()).isAir() && !belowState.isAir();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos neighborPos = pos.relative(direction);
             BlockState neighborState = level.getBlockState(neighborPos);
             if (neighborState.is(this)) {
diff --git a/net/minecraft/world/level/block/MossyCarpetBlock.java b/net/minecraft/world/level/block/MossyCarpetBlock.java
index dc2d53283c20be6ac31bdae82d8060d332f520d2..62039908276abd9a5d25517bd44974d53d4a2460 100644
--- a/net/minecraft/world/level/block/MossyCarpetBlock.java
+++ b/net/minecraft/world/level/block/MossyCarpetBlock.java
@@ -122,7 +122,7 @@ public class MossyCarpetBlock extends Block implements BonemealableBlock {
         BlockState belowState = null;
         createSides |= state.getValue(BASE);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             EnumProperty<WallSide> property = getPropertyForFace(direction);
             WallSide side = canSupportAtFace(level, pos, direction) ? (createSides ? WallSide.LOW : state.getValue(property)) : WallSide.NONE;
             if (side == WallSide.LOW) {
@@ -187,7 +187,7 @@ public class MossyCarpetBlock extends Block implements BonemealableBlock {
             BlockState noCarpetBaseState = Blocks.PALE_MOSS_CARPET.defaultBlockState().setValue(BASE, false);
             BlockState aboveState = getUpdatedState(noCarpetBaseState, level, pos.above(), true);
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 EnumProperty<WallSide> property = getPropertyForFace(direction);
                 if (aboveState.getValue(property) != WallSide.NONE && !sideSurvivalTest.getAsBoolean()) {
                     aboveState = aboveState.setValue(property, WallSide.NONE);
diff --git a/net/minecraft/world/level/block/RailState.java b/net/minecraft/world/level/block/RailState.java
index af5e9e8d26ad0d9d9580aa327d9f145ee3cb747e..ac2499dd591f56ed9945fa8155d9b6a02bc016f3 100644
--- a/net/minecraft/world/level/block/RailState.java
+++ b/net/minecraft/world/level/block/RailState.java
@@ -133,7 +133,7 @@ public class RailState {
     protected int countPotentialConnections() {
         int count = 0;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (this.hasRail(this.pos.relative(direction))) {
                 count++;
             }
diff --git a/net/minecraft/world/level/block/RedstoneWireBlock.java b/net/minecraft/world/level/block/RedstoneWireBlock.java
index 78be705043916e80a542ca0991a58f55780c6776..a74fe8ec1c6e48d202eeabf420243ef4cc8b8956 100644
--- a/net/minecraft/world/level/block/RedstoneWireBlock.java
+++ b/net/minecraft/world/level/block/RedstoneWireBlock.java
@@ -151,7 +151,7 @@ public class RedstoneWireBlock extends Block {
     private BlockState getMissingConnections(final BlockGetter level, BlockState state, final BlockPos pos) {
         boolean canConnectUp = !level.getBlockState(pos.above()).isRedstoneConductor(level, pos);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (!state.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected()) {
                 RedstoneSide sideConnection = this.getConnectingSide(level, pos, direction, canConnectUp);
                 state = state.setValue(PROPERTY_BY_DIRECTION.get(direction), sideConnection);
@@ -208,7 +208,7 @@ public class RedstoneWireBlock extends Block {
     ) {
         BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             RedstoneSide value = state.getValue(PROPERTY_BY_DIRECTION.get(direction));
             if (value != RedstoneSide.NONE && !level.getBlockState(blockPos.setWithOffset(pos, direction)).is(this)) {
                 blockPos.move(Direction.DOWN);
@@ -376,7 +376,7 @@ public class RedstoneWireBlock extends Block {
             }
             // Paper end - optimize redstone
 
-            for (Direction direction : Direction.Plane.VERTICAL) {
+            for (Direction direction : Direction.Plane.VERTICAL.faces) { // Pluto - Expose Direction$Plane's faces
                 level.updateNeighborsAt(pos.relative(direction), this);
             }
 
@@ -403,11 +403,11 @@ public class RedstoneWireBlock extends Block {
     }
 
     private void updateNeighborsOfNeighboringWires(final Level level, final BlockPos pos) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             this.checkCornerChangeAt(level, pos.relative(direction));
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos target = pos.relative(direction);
             if (level.getBlockState(target).isRedstoneConductor(level, target)) {
                 this.checkCornerChangeAt(level, target.above());
@@ -523,7 +523,7 @@ public class RedstoneWireBlock extends Block {
     public void animateTick(final BlockState state, final Level level, final BlockPos pos, final RandomSource random) {
         int power = state.getValue(POWER);
         if (power != 0) {
-            for (Direction horizontal : Direction.Plane.HORIZONTAL) {
+            for (Direction horizontal : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 RedstoneSide connection = state.getValue(PROPERTY_BY_DIRECTION.get(horizontal));
                 switch (connection) {
                     case UP:
@@ -600,7 +600,7 @@ public class RedstoneWireBlock extends Block {
     private void updatesOnShapeChange(final Level level, final BlockPos pos, final BlockState oldState, final BlockState newState) {
         Orientation orientation = ExperimentalRedstoneUtils.initialOrientation(level, null, Direction.UP);
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos relativePos = pos.relative(direction);
             if (oldState.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected() != newState.getValue(PROPERTY_BY_DIRECTION.get(direction)).isConnected()
                 && level.getBlockState(relativePos).isRedstoneConductor(level, relativePos)) {
diff --git a/net/minecraft/world/level/block/ScaffoldingBlock.java b/net/minecraft/world/level/block/ScaffoldingBlock.java
index 92bc27c26e18472ee2c8555b17fc5b5e05db4ce4..e678e14a475f9fd3d6a5f4545de1b60570918375 100644
--- a/net/minecraft/world/level/block/ScaffoldingBlock.java
+++ b/net/minecraft/world/level/block/ScaffoldingBlock.java
@@ -158,7 +158,7 @@ public class ScaffoldingBlock extends Block implements SimpleWaterloggedBlock {
             return 0;
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockState relativeState = level.getBlockState(relativePos.setWithOffset(pos, direction));
             if (relativeState.is(Blocks.SCAFFOLDING)) {
                 distance = Math.min(distance, relativeState.getValue(DISTANCE) + 1);
diff --git a/net/minecraft/world/level/block/SugarCaneBlock.java b/net/minecraft/world/level/block/SugarCaneBlock.java
index 4ad1680815087d753adafeeabb393fe37db72aeb..2eb5e4e4f1fd39202554ad0c81c6a250828367c9 100644
--- a/net/minecraft/world/level/block/SugarCaneBlock.java
+++ b/net/minecraft/world/level/block/SugarCaneBlock.java
@@ -89,7 +89,7 @@ public class SugarCaneBlock extends Block implements BonemealableBlock { // Purp
         if (stateBelow.is(BlockTags.SUPPORTS_SUGAR_CANE)) {
             BlockPos below = pos.below();
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 BlockState blockState = level.getBlockState(below.relative(direction));
                 FluidState fluidState = level.getFluidState(below.relative(direction));
                 if (fluidState.is(FluidTags.SUPPORTS_SUGAR_CANE_ADJACENTLY) || blockState.is(BlockTags.SUPPORTS_SUGAR_CANE_ADJACENTLY)) {
diff --git a/net/minecraft/world/level/block/VineBlock.java b/net/minecraft/world/level/block/VineBlock.java
index 81f87a0054cc2882679b9fade760d1f5a0772312..ace074e4100785fb72e3b25213b4e9f09b7628db 100644
--- a/net/minecraft/world/level/block/VineBlock.java
+++ b/net/minecraft/world/level/block/VineBlock.java
@@ -122,7 +122,7 @@ public class VineBlock extends Block {
 
         BlockState aboveState = null;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BooleanProperty property = getPropertyForFace(direction);
             if (state.getValue(property)) {
                 boolean canSupport = this.canSupportAtFace(level, pos, direction);
@@ -214,7 +214,7 @@ public class VineBlock extends Block {
 
                             BlockState aboveState = state;
 
-                            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                 if (random.nextBoolean() || !isAcceptableNeighbour(level, abovePos.relative(direction), direction)) {
                                     aboveState = aboveState.setValue(getPropertyForFace(direction), false);
                                 }
@@ -245,7 +245,7 @@ public class VineBlock extends Block {
     }
 
     private BlockState copyRandomFaces(final BlockState from, BlockState to, final RandomSource random) {
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (random.nextBoolean()) {
                 BooleanProperty propertyForFace = getPropertyForFace(direction);
                 if (from.getValue(propertyForFace)) {
diff --git a/net/minecraft/world/level/block/sounds/AmbientDesertBlockSoundsPlayer.java b/net/minecraft/world/level/block/sounds/AmbientDesertBlockSoundsPlayer.java
index 3bc75f501d0050c1a5b8cd0321369e251b4e9efa..d7436c11c7d1ea3102cbd0f2d09356106a3397ba 100644
--- a/net/minecraft/world/level/block/sounds/AmbientDesertBlockSoundsPlayer.java
+++ b/net/minecraft/world/level/block/sounds/AmbientDesertBlockSoundsPlayer.java
@@ -58,7 +58,7 @@ public class AmbientDesertBlockSoundsPlayer {
         int sidesChecked = 0;
         BlockPos.MutableBlockPos mutablePos = pos.mutable();
 
-        for (Direction dir : Direction.Plane.HORIZONTAL) {
+        for (Direction dir : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             mutablePos.set(pos).move(dir, 8);
             if (columnContainsTriggeringBlock(level, mutablePos) && matchingBlocksFound++ >= 3) {
                 return true;
diff --git a/net/minecraft/world/level/dimension/end/EnderDragonFight.java b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
index 03981d3c52a4dc735707efb80a8c0bfd37e64d90..1f55f2a3da0be41f8a04f268a1304bb33530ef04 100644
--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -596,7 +596,7 @@ public class EnderDragonFight extends SavedData {
             List<EndCrystal> crystals = Lists.newArrayList();
             BlockPos center = location.above(1);
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 List<EndCrystal> found = this.level.getEntitiesOfClass(EndCrystal.class, new AABB(center.relative(direction, 3)));
                 if (found.isEmpty()) {
                     return false; // CraftBukkit - return value
diff --git a/net/minecraft/world/level/levelgen/blending/Blender.java b/net/minecraft/world/level/levelgen/blending/Blender.java
index b7722239afbc04f77445584f5019fedc3c8f3ced..4ee5c165f4eff9e1a9830d4126d1e8efe063787d 100644
--- a/net/minecraft/world/level/levelgen/blending/Blender.java
+++ b/net/minecraft/world/level/levelgen/blending/Blender.java
@@ -322,7 +322,7 @@ public class Blender {
                     }
                 }
 
-                for (Direction direction : Direction.Plane.HORIZONTAL) {
+                for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                     if (region.getChunk(chunkPos.x() + direction.getStepX(), chunkPos.z() + direction.getStepZ()).isOldNoiseGeneration() != oldNoiseGeneration) {
                         int minX = direction == Direction.EAST ? 15 : 0;
                         int maxX = direction == Direction.WEST ? 0 : 15;
diff --git a/net/minecraft/world/level/levelgen/feature/BonusChestFeature.java b/net/minecraft/world/level/levelgen/feature/BonusChestFeature.java
index 6c8e3bfc589e29ea910f7c40f22414ceb3096c76..cedfff456b527e249560f36effa075d9efd759b8 100644
--- a/net/minecraft/world/level/levelgen/feature/BonusChestFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/BonusChestFeature.java
@@ -41,7 +41,7 @@ public record BonusChestFeature() implements Feature {
                     RandomizableContainer.setBlockEntityLootTable(level, random, chestPos, BuiltInLootTables.SPAWN_BONUS_CHEST);
                     BlockState torch = Blocks.TORCH.defaultBlockState();
 
-                    for (Direction direction : Direction.Plane.HORIZONTAL) {
+                    for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                         BlockPos torchPos = chestPos.relative(direction);
                         if (torch.canSurvive(level, torchPos)) {
                             level.setBlock(torchPos, torch, Block.UPDATE_CLIENTS);
diff --git a/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java b/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java
index 20e2d649ad1f7cca5ab807f2afa5a787d378cfb2..d6444de6bd96f15c0cbf0c8c1a1dd063bdcf582b 100644
--- a/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/EndPodiumFeature.java
@@ -65,7 +65,7 @@ public record EndPodiumFeature(boolean active) implements Feature {
 
         BlockPos centerOfPillar = origin.above(2);
 
-        for (Direction face : Direction.Plane.HORIZONTAL) {
+        for (Direction face : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             this.setBlock(level, centerOfPillar.relative(face), Blocks.WALL_TORCH.defaultBlockState().setValue(WallTorchBlock.FACING, face));
         }
 
diff --git a/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java b/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
index c7f2cb020d35618b9521149317e0a3b69f2e7601..223f1f998709dfd31c626e488076fb8fc678bd24 100644
--- a/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/MonsterRoomFeature.java
@@ -100,7 +100,7 @@ public record MonsterRoomFeature() implements Feature {
                     if (level.isEmptyBlock(chestPos)) {
                         int wallCount = 0;
 
-                        for (Direction direction : Direction.Plane.HORIZONTAL) {
+                        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                             if (level.getBlockState(chestPos.relative(direction)).isSolid()) {
                                 wallCount++;
                             }
diff --git a/net/minecraft/world/level/levelgen/feature/MultifaceGrowthFeature.java b/net/minecraft/world/level/levelgen/feature/MultifaceGrowthFeature.java
index c1bd1fa54bf0d033f0a30b58f192e889289ee728..d937b0a42b7f5da154da496edde331dd3ab7e0ce 100644
--- a/net/minecraft/world/level/levelgen/feature/MultifaceGrowthFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/MultifaceGrowthFeature.java
@@ -135,7 +135,7 @@ public record MultifaceGrowthFeature(
         }
 
         if (this.canPlaceOnWall) {
-            Direction.Plane.HORIZONTAL.forEach(validDirections::add);
+            validDirections.addElements(validDirections.size(), Direction.Plane.HORIZONTAL.faces); // Pluto - Expose Direction$Plane's faces
         }
 
         return validDirections;
diff --git a/net/minecraft/world/level/levelgen/feature/SpeleothemClusterFeature.java b/net/minecraft/world/level/levelgen/feature/SpeleothemClusterFeature.java
index 50a31e0fff0c46f11764190753416a9db855ccc8..cd4549629ceecd2d275733d329ea4e6d4cb5939a 100644
--- a/net/minecraft/world/level/levelgen/feature/SpeleothemClusterFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/SpeleothemClusterFeature.java
@@ -233,7 +233,7 @@ public record SpeleothemClusterFeature(
                 return false;
             }
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 if (!this.canBeAdjacentToWater(level, pos.relative(direction))) {
                     return false;
                 }
diff --git a/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java b/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
index 62a7aba42e178a8de1c79007fe10a4eea9bbe5a4..70d457f0024433e5cfcbcee0798fca0d1b841665 100644
--- a/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
+++ b/net/minecraft/world/level/levelgen/feature/UnderwaterMagmaFeature.java
@@ -68,7 +68,7 @@ public record UnderwaterMagmaFeature(int floorSearchRange, int placementRadiusAr
 
     private boolean isValidPlacement(final WorldGenLevel level, final BlockPos pos) {
         if (!isWaterOrAir(level.getBlockState(pos)) && !this.isVisibleFromOutside(level, pos.below(), Direction.UP)) {
-            for (Direction neighbourDir : Direction.Plane.HORIZONTAL) {
+            for (Direction neighbourDir : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 if (this.isVisibleFromOutside(level, pos.relative(neighbourDir), neighbourDir.getOpposite())) {
                     return false;
                 }
diff --git a/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java b/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
index e4bd972f63b5b2c90ad77fce8d1325f7f69589e4..64ce346d0e40d5436813af3b25abad88df10afd1 100644
--- a/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/foliageplacers/FoliagePlacer.java
@@ -131,7 +131,7 @@ public abstract class FoliagePlacer {
         BlockPos logPos = origin.below();
         BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos();
 
-        for (Direction alongEdge : Direction.Plane.HORIZONTAL) {
+        for (Direction alongEdge : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Direction toEdge = alongEdge.getClockWise();
             int offsetToEdge = toEdge.getAxisDirection() == Direction.AxisDirection.POSITIVE ? currentRadius + offset : currentRadius;
             pos.setWithOffset(origin, 0, y - 1, 0).move(toEdge, offsetToEdge).move(alongEdge, -currentRadius);
diff --git a/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java b/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java
index f6bf07da7d42325f17a90e01d69c01088dcdf8c0..31417587489106c49fe14fec737f01e0d8049451 100644
--- a/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java
+++ b/net/minecraft/world/level/levelgen/feature/rootplacers/MangroveRootPlacer.java
@@ -59,7 +59,7 @@ public class MangroveRootPlacer extends RootPlacer {
 
         rootPositions.add(trunkOrigin.below());
 
-        for (Direction dir : Direction.Plane.HORIZONTAL) {
+        for (Direction dir : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos pos = trunkOrigin.relative(dir);
             List<BlockPos> positionsInDirection = Lists.newArrayList();
             if (!this.simulateRoots(level, random, pos, dir, trunkOrigin, positionsInDirection, 0)) {
diff --git a/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java b/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
index 4fafa9253fecc4f5293d7ac2fe9caa983fe28556..e877197da6d471a75fed879216b89393f558c3da 100644
--- a/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
+++ b/net/minecraft/world/level/levelgen/feature/treedecorators/CocoaDecorator.java
@@ -33,7 +33,7 @@ public class CocoaDecorator extends TreeDecorator {
                     .filter(pos -> pos.getY() - treeY <= 2)
                     .forEach(
                         pos -> {
-                            for (Direction direction : Direction.Plane.HORIZONTAL) {
+                            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                 if (random.nextFloat() <= 0.25F) {
                                     Direction opposite = direction.getOpposite();
                                     BlockPos cocoaPos = pos.offset(opposite.getStepX(), 0, opposite.getStepZ());
diff --git a/net/minecraft/world/level/levelgen/structure/StructurePiece.java b/net/minecraft/world/level/levelgen/structure/StructurePiece.java
index dc17f0ab266a6370ffec5e42342a60710df39886..71acfeb0b9a4f1b3358e006705441484e14d6668 100644
--- a/net/minecraft/world/level/levelgen/structure/StructurePiece.java
+++ b/net/minecraft/world/level/levelgen/structure/StructurePiece.java
@@ -454,7 +454,7 @@ public abstract class StructurePiece {
     public static BlockState reorient(final BlockGetter level, final BlockPos blockPos, final BlockState blockState) {
         Direction solidNeighbor = null;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos relativePos = blockPos.relative(direction);
             BlockState state = level.getBlockState(relativePos);
             if (state.is(Blocks.CHEST)) {
diff --git a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
index cf3e1e74736ff9a8f51bfdd0132b3da83dd6820a..2f804b17eddf8884761a6e67c4a30b8e607e9a62 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/DesertPyramidPiece.java
@@ -292,7 +292,7 @@ public class DesertPyramidPiece extends ScatteredFeaturePiece {
             this.placeBlock(level, Blocks.CHISELED_SANDSTONE.defaultBlockState(), 10, -10, 13, chunkBB);
             this.placeBlock(level, Blocks.CUT_SANDSTONE.defaultBlockState(), 10, -11, 13, chunkBB);
 
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 if (!this.hasPlacedChest[direction.get2DDataValue()]) {
                     int xo = direction.getStepX() * 2;
                     int zo = direction.getStepZ() * 2;
diff --git a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
index eefc3c2965b69ab008005940b3d8e9516400bd31..c4e1c648c2520381365240d4c916675db6011c20 100644
--- a/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
+++ b/net/minecraft/world/level/levelgen/structure/structures/WoodlandMansionPieces.java
@@ -167,7 +167,7 @@ public class WoodlandMansionPieces {
         public @Nullable Direction get1x2RoomDirection(
             final WoodlandMansionPieces.SimpleGrid grid, final int x, final int y, final int floorNum, final int roomId
         ) {
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 if (this.isRoomId(grid, x + direction.getStepX(), y + direction.getStepZ(), floorNum, roomId)) {
                     return direction;
                 }
@@ -279,7 +279,7 @@ public class WoodlandMansionPieces {
 
                 List<Direction> potentialCorridors = new ArrayList<>();
 
-                for (Direction direction : Direction.Plane.HORIZONTAL) {
+                for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                     if (this.thirdFloorGrid.get(roomEndX + direction.getStepX(), roomEndY + direction.getStepZ()) == 0) {
                         potentialCorridors.add(direction);
                     }
@@ -547,7 +547,7 @@ public class WoodlandMansionPieces {
                                 && (roomData & WoodlandMansionPieces.MansionGrid.ROOM_CORRIDOR_FLAG) == WoodlandMansionPieces.MansionGrid.ROOM_CORRIDOR_FLAG;
                             doorDirs.clear();
                             if ((roomData & WoodlandMansionPieces.MansionGrid.ROOM_DOOR_FLAG) == WoodlandMansionPieces.MansionGrid.ROOM_DOOR_FLAG) {
-                                for (Direction direction : Direction.Plane.HORIZONTAL) {
+                                for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                                     if (grid.get(x + direction.getStepX(), y + direction.getStepZ()) == 1) {
                                         doorDirs.add(direction);
                                     }
diff --git a/net/minecraft/world/level/material/FlowingFluid.java b/net/minecraft/world/level/material/FlowingFluid.java
index dc81802f4475bd3cd3ed4e1f2cf13db478337e60..eed24d905e24b4107352ed3e1392c63ad6fb44e3 100644
--- a/net/minecraft/world/level/material/FlowingFluid.java
+++ b/net/minecraft/world/level/material/FlowingFluid.java
@@ -98,7 +98,7 @@ public abstract class FlowingFluid extends Fluid {
         double flowZ = 0.0;
         BlockPos.MutableBlockPos blockPos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             blockPos.setWithOffset(pos, direction);
             FluidState neighbourFluid = level.getFluidState(blockPos);
             if (this.affectsFlow(neighbourFluid)) {
@@ -128,7 +128,7 @@ public abstract class FlowingFluid extends Fluid {
 
         Vec3 flow = new Vec3(flowX, 0.0, flowZ);
         if (fluidState.getValue(FALLING)) {
-            for (Direction direction : Direction.Plane.HORIZONTAL) {
+            for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 blockPos.setWithOffset(pos, direction);
                 if (this.isSolidFace(level, blockPos, direction) || this.isSolidFace(level, blockPos.above(), direction)) {
                     flow = flow.normalize().add(0.0, -6.0, 0.0);
@@ -219,7 +219,7 @@ public abstract class FlowingFluid extends Fluid {
         int neighbourSources = 0;
         BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos relativePos = mutablePos.setWithOffset(pos, direction);
             BlockState blockState = level.getBlockStateIfLoaded(relativePos); // Paper - Prevent chunk loading from fluid flowing
             if (blockState == null) continue; // Paper - Prevent chunk loading from fluid flowing
@@ -351,7 +351,7 @@ public abstract class FlowingFluid extends Fluid {
         visited.clear();
         queue.clear();
 
-        for (Direction dir : Direction.Plane.HORIZONTAL) {
+        for (Direction dir : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             if (dir == from) continue;
 
             BlockPos neighborPos = pos.relative(dir); // immutable
@@ -381,7 +381,7 @@ public abstract class FlowingFluid extends Fluid {
 
             if (current.depth >= slopeFindDistance) continue;
 
-            for (Direction dir : Direction.Plane.HORIZONTAL) {
+            for (Direction dir : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
                 if (dir == current.excludedDir) continue;
 
                 BlockPos nextPos = current.pos.relative(dir); // immutable
@@ -507,7 +507,7 @@ public abstract class FlowingFluid extends Fluid {
     private int sourceNeighborCount(final LevelReader level, final BlockPos pos) {
         int count = 0;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos testPos = pos.relative(direction);
             FluidState testFluidState = level.getFluidState(testPos);
             if (this.isSourceBlockOfThisType(testFluidState)) {
@@ -523,7 +523,7 @@ public abstract class FlowingFluid extends Fluid {
         Map<Direction, FluidState> result = Maps.newEnumMap(Direction.class);
         FlowingFluid.SpreadContext context = null;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos testPos = pos.relative(direction);
             BlockState testState = level.getBlockStateIfLoaded(testPos); // Paper - Prevent chunk loading from fluid flowing
             if (testState == null) continue; // Paper - Prevent chunk loading from fluid flowing
diff --git a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
index 49d959160337f61a40cb94c31aec8dad0aee3839..a4fa389670f3f4ac6df5d2d95aa148843fb7630d 100644
--- a/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/SwimNodeEvaluator.java
@@ -59,7 +59,7 @@ public class SwimNodeEvaluator extends NodeEvaluator {
             }
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Direction secondDirection = direction.getClockWise();
             if (hasMalus(nodes.get(direction)) && hasMalus(nodes.get(secondDirection))) {
                 Node diagonalNode = this.findAcceptedNode(
diff --git a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
index 8826c5c7fc9dc77d3381551e43c25eca84a198e9..aeaf9188a521eb33b180ea7871265e37b79ce55d 100644
--- a/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
+++ b/net/minecraft/world/level/pathfinder/WalkNodeEvaluator.java
@@ -127,7 +127,7 @@ public class WalkNodeEvaluator extends NodeEvaluator {
 
         double posHeight = this.getFloorLevel(new BlockPos(pos.x, pos.y, pos.z));
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Node node = this.findAcceptedNode(
                 pos.x + direction.getStepX(), pos.y, pos.z + direction.getStepZ(), jumpSize, posHeight, direction, blockPathTypeCurrent
             );
@@ -137,7 +137,7 @@ public class WalkNodeEvaluator extends NodeEvaluator {
             }
         }
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             Direction secondDirection = direction.getClockWise();
             if (this.isDiagonalValid(pos, this.reusableNeighbors[direction.get2DDataValue()], this.reusableNeighbors[secondDirection.get2DDataValue()])) {
                 Node diagonalNode = this.findAcceptedNode(
diff --git a/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java b/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
index 01a3662cb6539b842e5d62ba150cc0f5990caa04..da893a2d1c6e60deb7cbfd959264d8d072bf440b 100644
--- a/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
+++ b/net/minecraft/world/level/redstone/RedstoneWireEvaluator.java
@@ -29,7 +29,7 @@ public abstract class RedstoneWireEvaluator {
     protected int getIncomingWireSignal(final Level level, final BlockPos pos) {
         int wireSignal = 0;
 
-        for (Direction direction : Direction.Plane.HORIZONTAL) {
+        for (Direction direction : Direction.Plane.HORIZONTAL.faces) { // Pluto - Expose Direction$Plane's faces
             BlockPos neighborPos = pos.relative(direction);
             BlockState neighborState = level.getBlockState(neighborPos);
             wireSignal = Math.max(wireSignal, this.getWireSignal(neighborPos, neighborState));
